cleverhans/attacks.py

```diff
@@ -451,6 +451,154 @@ def parse_params(self, eps=0.3, eps_iter=0.05, nb_iter=10, y=None,
+                                'y': np.float32,
+                                'y_target': np.float32,
+                                'clip_min': np.float32,
+                                'clip_max': np.float32}
+        self.structural_kwargs = ['ord', 'nb_iter', 'decay_factor']
+
+        if not isinstance(self.model, Model):
+            self.model = CallableModelWrapper(self.model, 'probs')
+
+    def generate(self, x, **kwargs):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        :param x: The model's symbolic inputs.
+        :param eps: (required float) maximum distortion of adversarial example
+                    compared to original input
+        :param eps_iter: (required float) step size for each attack iteration
+        :param nb_iter: (required int) Number of attack iterations.
+        :param y: (optional) A tensor with the model labels.
+        :param y_target: (optional) A tensor with the labels to target. Leave
+                         y_target=None if y is also set. Labels should be
+                         one-hot-encoded.
+        :param ord: (optional) Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param decay_factor: (optional) Decay factor for the momentum term.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        import tensorflow as tf
+
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+
+        # Initialize loop variables
+        momentum = 0
+        adv_x = x
+
+        # Fix labels to the first model predictions for loss computation
+        y, nb_classes = self.get_or_guess_labels(x, kwargs)
+        y = y / tf.reduce_sum(y, 1, keep_dims=True)
+        targeted = (self.y_target is not None)
+
+        from . import utils_tf
+        for i in range(self.nb_iter):
+            # Compute loss
+            preds = self.model.get_probs(adv_x)
+            loss = utils_tf.model_loss(y, preds, mean=False)
+            if targeted:
+                loss = -loss
+
+            # Define gradient of loss wrt input
+            grad, = tf.gradients(loss, adv_x)
+
+            # Normalize current gradient and add it to the accumulated gradient
+            red_ind = list(xrange(1, len(grad.get_shape())))
+            grad = grad / tf.reduce_mean(tf.abs(grad),
+                                         reduction_indices=red_ind,
+                                         keep_dims=True)
+            momentum = self.decay_factor * momentum + grad
+
+            if self.ord == np.inf:
+                normalized_grad = tf.sign(momentum)
+            elif self.ord in [1, 2]:
+                reduc_ind = list(xrange(1, len(momentum.get_shape())))
+                if self.ord == 1:
+                    norm = tf.reduce_sum(tf.abs(momentum),
```

Comment:
It would be nice to apply the same logic that @goodfeli applied to the clipping function in order to avoid numerical instabilities in the norm: see https://github.com/tensorflow/cleverhans/blob/master/cleverhans/utils_tf.py#L417-L431

Reply:
Alternatively, you may be able to reuse the same function `clip_eta` to improve maintainability

Reply:
Ok, I'll address it. It's actually the same code as fgm https://github.com/tensorflow/cleverhans/blob/master/cleverhans/attacks_tf.py#L65-L87.
I reuse the 'clip_eta' after adding the perturbation.